src/oemof/solph/components/_generic_storage.py

```diff
@@ -1232,24 +1233,75 @@ def _old_storage_capacity_rule_end(block):
                             " Value for {} is missing.".format(n)
                         )
                         raise ValueError(msg)
+                    # get the period matrix describing the temporal distance
+                    # between all period combinations. Row indexes indicating
+                    # the investment period, column indexes the decommissioning
+                    # period.
+                    periods_matrix = m.es.periods_matrix
+                    # matrix = np.where(matrix == 0, np.nan, matrix)
+
+                    # get the index of the minimum value in each row greater
+                    # equal than the lifetime. This value equals the
+                    # decommissioning period if not zero. The index of this
+                    # value represents the investment period.
+                    decomm_periods = np.argmin(
+                        np.where(
+                            (periods_matrix >= lifetime),
+                            periods_matrix,
+                            np.inf,
+                        ),
+                        axis=1,
+                    )
+
+                    # first period doesn't have any decommissioning
+                    expr = self.old_end[n, 0] == 0
+                    self.old_rule_end.add((n, 0), expr)
+
+                    # all periods not in decomm_periods have no decommissioning
                     for p in m.PERIODS:
-                        # No shutdown in first period
-                        if p == 0:
+                        if p not in decomm_periods and p != 0:
                             expr = self.old_end[n, p] == 0
                             self.old_rule_end.add((n, p), expr)
```

Comment:
You could combine the 0th period into that statement, couldn't you?

I.e.

```python3
if p not in decomm_periods or p == 0:
    ...
```

Reply:
No, this doesn't work as the `np.argmin(np.where..)` returns 0 if the condition is not matched. This would lead to having the 0th period constraint multiple times.